src/transformers/pipelines/fill_in_middle.py

```diff
@@ -0,0 +1,333 @@
+    "codellama/CodeLlama-13b-hf": ("▁<PRE>", "▁<MID>", "▁<SUF>"),
+    "codellama/CodeLlama-34b-hf": ("▁<PRE>", "▁<MID>", "▁<SUF>"),
+    "codellama/CodeLlama-7b-Python-hf": ("▁<PRE>", "▁<MID>", "▁<SUF>"),
+    "codellama/CodeLlama-13b-Python-hf": ("▁<PRE>", "▁<MID>", "▁<SUF>"),
+    "codellama/CodeLlama-34b-Python-hf": ("▁<PRE>", "▁<MID>", "▁<SUF>"),
+}
+
+
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class FimPipeline(Pipeline):
+    """
+    Fill-in-the-middle Language generation pipeline. This pipeline predicts the middle part of a text given it's prefix and suffix
+
+    Example:
+    ```python
+    >>> from transformers import pipeline
+    >>> PROMPT = '''
+        def fib(x: int) -> int:
+            <FILL_ME>
+            return fib(x-1) + fib(x-2)
+    '''
+    >>> generator = pipeline(model="codellama/CodeLlama-7b-hf")
+    >>> print(generator(PROMPT, do_sample=False))
+    [{'generated_text': "\ndef fib(x: int) -> int:\n\tif x == 0:\n\t\treturn 0\n\tif x == 1:\n\t\treturn 1\n\telse:\n\t\treturn fib(x-1) + fib(x-2)\n"}]
+    ```
+
+    Learn more about the basics of using a pipeline in the [pipeline tutorial](../pipeline_tutorial). You can pass text
+    generation parameters to this pipeline to control stopping criteria, decoding strategy, and more. Learn more about
+    text generation parameters in [Text generation strategies](../generation_strategies) and [Text
+    generation](text_generation).
+    This language generation pipeline can currently be loaded from [`pipeline`] using the following task identifier:
+    `"fill-in-middle"`.
+    The models that this pipeline can use are models that have been trained with an autoregressive language modeling
+    objective, which includes the uni-directional models in the library (e.g. gpt2). See the list of available models
+    on [huggingface.co/models](https://huggingface.co/models?filter=text-generation).
+    """
+
+    # There are two modes in which infilling is supported: PSM (Prefix-Suffix-Middle) and SPM (Suffix-Prefix-Middle)
+    # CodeLlama's HF implementation has a third mode called 'suffix-first' which is added here for total compatibility with CodeLlama family
+    # The placement of prefix and suffix along with their respective sentinel tokens depends on the mode
+    # More information on this: https://arxiv.org/abs/2207.14255
+    DEFAULT_INFILL_MODE = "psm"
+    DEFAULT_INFILL_TOKEN = "<FILL_ME>"
+    ALL_INFILL_MODES = ["psm", "spm", "suffix-first"]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Check if the provided model is among the supported ones
+        if self.model.name_or_path not in SUPPORTED_MODELS:
+            raise ValueError(f"The model '{self.model.name_or_path}' is not supported for the Fill-in-Middle task.")
+
+        # Get the Prefix, Suffix and Middle tokens for the corresponding model
+        self.PREFIX_TOKEN, self.MIDDLE_TOKEN, self.SUFFIX_TOKEN = SUPPORTED_MODELS[self.model.name_or_path]
```

Comment:
that is not super super scalable, if someone has a custome model he also needs to register these variables. We could add them to the generation config IMO

Reply:
@ArthurZucker For this to work, we need to first add the infill tokens to GenerationConfigs for all Codellama models and the Starcoder model right?

Reply:
Mmm they should be in the tokenizer_config.json.
We can just save the tokens by doing `CodeLlamaTokenizer.from_pretrained("...", prefix_token = ""). It will be saved. If they are not in the tokenizer then error out.